Add support for Continuous "Low Frequency Generators"

tknopp opened the pull request:
This is my proposal to address #87.

* numPeriods in its current form have been removed. They were kind of ambiguous with numPatches
* numPatches is renamed to numPeriods

**Continuous "Low Frequency Generators"** 

The field `offsetField` may have a finer resolution then the number of Periods J. The number of "Low Frequency Generator" steps is called Y and usually Y will be a multiple of J. We may even want to enforce that since this would allow to reason about the field at the beginning of each drive-field cycle

**How to specify "Low Frequency Generators"**
There are different ways to specify the low frequency fields. We need different forms to all also describing scanners like the Clinical Demonstrator (CD) from Philips
* On the highest level one specifies `gradient` and `fieldFreePoint`. This requires precise knowledge on the location of the FFP.
* One level down: Specify `gradient` and `offsetField`.
* Lowest Level: On the lowest level it does not make sense to specify the gradient anymore. Therefore one will simply specify the progression of the field generators over time. This can either be done by specifying the field (which requires an additional reference point in the future) or the current. In the CD the relationsship is strongly nonlinear

I introduced a new group `/acquisition/generator` to describe the field generators. In the future we will want to store field profiles here.

hofmannmartin commented on specification/MDF.tex:
```diff
@@ -280,24 +280,45 @@ \subsection{Acquisition Parameters (group: \inlvar{/acquisition/}, non-optional)
-\inlvar{gradient} & \inltab{Float64} & $J \times 3$ & Tm$^{-1}\mu_0^{-1}$ & yes & Gradient strength of the selection field in $x$, $y$, and $z$ directions \\ \hline
-\inlvar{offsetField} & \inltab{Float64} & $J \times 3$ & T$\mu_0^{-1}$ & yes & Offset field applied for each patch in the measurement sequence \\ \hline
-\inlvar{offsetFieldShift} & \inltab{Float64} & $J \times 3$ & m & yes & Position of the field free point (relative to origin/center) \\ \hline
+\inlvar{gradient} & \inltab{Float64} & $Y \times 3$ & Tm$^{-1}\mu_0^{-1}$ & yes & Gradient strength of the selection field in $x$, $y$, and $z$ directions \\ \hline
```
Do we really want to have one Y?

Is it not better to split into Y_g (gradient) and Y_o (offset)? Maybe we want to manipulate both at different time scales.

tknopp replied:
could make sense. Although this gives a lot a freedom. In that case one could also have Y_d for the drive field.

hofmannmartin replied:
but the drive field can be handled via the custom wave form

hofmannmartin replied:
In my opinion all this low frequency stuff should only affect the signal encoding and not the signal generation in MPI. Good question, how a low freq modulation of df amplitude should be handled. For low enough frequencies this could work.

tknopp replied:
I remove the DF custom waveform since this was kind of ambiguous.

hofmannmartin replied:
I see

tknopp replied:
It was not only ambiguous but very inflexible. You either had to use the custom waveform for all channels or for non.

tknopp replied:
But back to topic: We could also not introduce `Y_g` and `Y_o`. This is a little bit more strict and the worst that can happen is that the amount of metadata is a little higher. If we really feel that this is an issue, we can introduce `Y_g` and `Y_o` without breaking the spec. The case `Y_g` = `Y_o` would also be handled.

hofmannmartin replied:
In this case lets stick with one `Y`.

hofmannmartin commented on specification/MDF.tex:
```diff
@@ -280,24 +280,45 @@ \subsection{Acquisition Parameters (group: \inlvar{/acquisition/}, non-optional)
+In MPI, a frame groups all data used to reconstruct a single MPI image/tomogram. In the simplest scenario, this data is acquired during one \inlvar{/drivefield/period}. The acquisition time increases by a factor of \inlvar{numAverages} if averaging is applied and by a factor of \inlvar{numPeriods} if data is recorded over several (averaged) cycles. 
+%In a multi-patch setting $J$ \inlvar{offsetField}s or mechanical movements shift the gradient field by \inlvar{offsetFieldShift}, where $J$ is the number of patches \inlvar{numPatches} of a multi-patch measurement. For each shift at least one full drive field cycle is used to acquire the measurement data. %For instance, a Cartesian 2D trajectory with 100 lines could be realized by setting \mbox{ ...
+
+\textcolor{red}{TODO: Discuss the dimension $Y$. It is usually equal to $J$ when considering a static multi-patch scenario. But it can also be finer resolved to describe continuous FF.}
```
We need to be able to have specify Y values within one drive field period with the first and last value exactly at the start and end end of a DF cycle (since we might not know what happens in between patches). The values in between could be equidistantly spread out. If we define it that way we can have static focus fields with Y=1,since the one value will be first and last at the same time.

tknopp replied:
Yes I think we have that if `Y` is a multiple of `J`. Then the first and last value will always be on the boundary. I think we should enforce this.

hofmannmartin replied:
Agree if we include 
> one value will be first and last at the same time.

into the specification

tknopp replied:
yes lets do that.

hofmannmartin replied:
give me a ping if you have finished an update on this one. I will look into it.

hofmannmartin commented on specification/MDF.tex:
```diff
@@ -319,7 +336,6 @@ \subsubsection{Drive Field (group: \inlvar{/acquisition/drivefield/}, non-option
 \inlvar{strength} & \inltab{Float64} & $J \times D \times F $ & T$\mu_0^{-1}$ & no & Applied drive field strength \\ \hline
 \inlvar{phase} & \inltab{Float64} & $J \times D \times F$ & rad, $[-\pi,\pi)$ & no & Applied drive field phase $\varphi$\\ \hline
 \inlvar{baseFrequency} & \inltab{Float64} & 1 & Hz & no & Base frequency to derive drive field frequencies \\ \hline
-\inlvar{customWaveform} & \inltab{Float64} & $D \times F \times U$ & 1 & yes & Custom waveform table \\ \hline
```
Why did this went away?

tknopp commented:
I have revised this branch. Added the requested description on `Y`. Further after discussion with @SzwarP I have changed the name `numPeriods` to `numPeriodsPerFrame`. This makes crystal clear what is meant. We will then have `numPeriods = numPeriodsPerFrame * numFrames`. Will merge now. Refinements can be made on master.
